internal/zinc-core/src/main/scala/sbt/internal/inc/Stamp.scala

```diff
@@ -141,8 +127,25 @@ object Stamper {
     catch { case i: IOException => EmptyStamp }
   }
 
-  val forHash = (toStamp: File) => tryStamp(Hash.ofFile(toStamp))
-  val forLastModified = (toStamp: File) => tryStamp(new LastModified(toStamp.lastModified()))
+  private final val xxHash = LongHashFunction.xx()
+  private final val BufferSize = 8192
+  final val forHash: (File => Stamp) = (toStamp: File) => {
+    if (!toStamp.exists() || toStamp.isDirectory) EmptyStamp
+    else {
+      sbt.io.Using.fileInputStream(toStamp) { (is: InputStream) =>
+        val acc = new Array[Long](2)
+        val buffer = new Array[Byte](BufferSize)
+        while (is.read(buffer) >= 0) {
+          val checksumChunk = xxHash.hashBytes(buffer)
```

Comment:
Can you include a reference to where you found this pattern?

Alternatively, using a hash library with a more usable interface would be great. Guava's is particularly good, and includes [hash of file](http://google.github.io/guava/releases/19.0/api/docs/com/google/common/io/Files.html#hash(java.io.File,%20com.google.common.hash.HashFunction)).